File: mvvm/src/main/java/com/example/mvvm/viewmodel/LoginViewModel.java
package com.example.mvvm.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.mvvm.bean.User;
import com.example.mvvm.model.LoginModel;
import com.example.mvvm.state.ResultState;

public class LoginViewModel {
    private MutableLiveData<ResultState<User>> loginLiveData;
    private LoginModel loginModel;

    public LoginViewModel() {
        loginLiveData = new MutableLiveData<>();
        loginModel = new LoginModel();
    }

    public LiveData<ResultState<User>> getLoginLiveData() {
        return loginLiveData;
    }

    public void login(String username, String password) {
        loginLiveData.postValue(new ResultState<>(ResultState.LOADING));
        loginModel.login(username, password, new LoginModel.OnLoginCallback() {
            @Override
            public void onLoginSuccess(User user) {
                loginLiveData.postValue(new ResultState<>(ResultState.SUCCESS, user));
            }

            @Override
            public void onLoginError(int errCode, String errMsg) {
                loginLiveData.postValue(new ResultState<>(ResultState.ERROR, errCode, errMsg));
            }
        });
    }
}

File: mvvm_jetpack/src/main/java/com/example/mvvm_jetpack/ui/LoginViewModel.kt
package com.example.mvvm_jetpack.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.common.utils.livedata.SingleLiveData
import com.example.mvvm_jetpack.data.model.bean.User
import com.example.mvvm_jetpack.data.repository.LoginRepository
import com.example.mvvm_jetpack.data.state.ResultState
import com.example.mvvm_jetpack.ui.state.UiState
import kotlinx.coroutines.launch

class LoginViewModel : ViewModel() {
    private val loginRepo: LoginRepository by lazy { LoginRepository() }
    private val _loginState = SingleLiveData<UiState<User>>()
    val loginState: LiveData<UiState<User>> = _loginState

    fun login(username: String, password: String) {
        _loginState.value = UiState.Loading
        viewModelScope.launch {
            when (val resultState = loginRepo.login(username, password)) {
                is ResultState.Success -> {
                    _loginState.value = UiState.Success(resultState.data)
                }
                is ResultState.Error -> {
                    _loginState.value = UiState.Error(resultState.errCode, resultState.errMsg)
                }
            }
        }
    }
}